uom/common/container/kdtree.h

```diff
@@ -0,0 +1,627 @@
+
+  auto collect_from_child = [this](TreeNode* child) {
+    child->parent = this;
+    this->size += child->size;
+    this->range.extend(child->range);
+    this->num_invalid_nodes += child->num_invalid_nodes;
+  };
+
+  if (left != nullptr) {
+    collect_from_child(left.get());
+  }
+  if (right != nullptr) {
+    collect_from_child(right.get());
+  }
+
+  this->radius = ((this->range.max() - this->range.min()) * 0.5).norm();
+  this->tree_deleted =
+      this->point_deleted && (!left || left->tree_deleted) && (!right || right->tree_deleted);
+}
+
+template <typename PointType>
+void KdTree<PointType>::TreeNode::PushDown() {
+  auto apply_to_child = [this](TreeNode* child) {
+    child->tree_deleted |= this->tree_deleted;
+    child->point_deleted |= child->tree_deleted;
+    if (this->tree_deleted) {
+      child->num_invalid_nodes = child->size;
+    }
+  };
+
+  if (left != nullptr) {
+    apply_to_child(left.get());
+  }
+  if (right != nullptr) {
+    apply_to_child(right.get());
+  }
+}
+
+template <typename PointType>
+double KdTree<PointType>::TreeNode::RangeTo(const PointType& query_point) const {
+  double squared_distance = 0;
+  for (int d = 0; d < kDim; ++d) {
+    const double val = query_point[d];
+    const double min_val = this->range.min()[d];
+    const double max_val = this->range.max()[d];
+    if (val < min_val) {
+      squared_distance += (val - min_val) * (val - min_val);
+    }
+    if (val > max_val) {
+      squared_distance += (val - max_val) * (val - max_val);
+    }
+  }
+  return std::sqrt(squared_distance);
+}
+
+template <typename PointType>
+void KdTree<PointType>::BuildTree(const std::vector<PointType>& points) {
+  if (points.empty()) {
+    return;
+  }
+
+  // Copy points as we need to sort the input points to build the tree.
+  std::vector<PointType> workspace = points;
+  auto workspace_span = absl::MakeSpan(workspace);
+  BuildTreeToNodeRecursive(workspace_span, &root_);
+}
+
+template <typename PointType>
+int KdTree<PointType>::FindBestSplitAxisAndSortByMaxSpan(absl::Span<PointType> points) {
+  CHECK_GT(points.size(), 0);
+
+  double min_value[kDim];  // NOLINT
+  double max_value[kDim];  // NOLINT
+  for (int d = 0; d < kDim; ++d) {
+    min_value[d] = points[0][d];
+    max_value[d] = points[0][d];
+  }
+  for (int i = 1; i < static_cast<int>(points.size()); ++i) {
+    for (int d = 0; d < kDim; ++d) {
+      min_value[d] = std::min(min_value[d], points[i][d]);
+      max_value[d] = std::max(max_value[d], points[i][d]);
+    }
+  }
+
+  int max_dim = 0;
+  double max_span = max_value[0] - min_value[0];
+  for (int d = 1; d < kDim; ++d) {
+    const double curr_span = max_value[d] - min_value[d];
+    if (curr_span > max_span) {
+      max_dim = d;
+      max_span = curr_span;
+    }
+  }
+
+  auto cmp = [max_dim](auto&& lhs, auto&& rhs) { return lhs[max_dim] < rhs[max_dim]; };
+  std::nth_element(points.begin(), points.begin() + points.size() / 2, points.end(), cmp);
```

Comment:
how about 2<2<3
the result will be 
   └──2
       ├──2
       └──3
which is not expected

Reply:
Are you suggesting adding duplicate point checks? This is indeed not the case in the current implementation, and it is assumed that all points do not overlap.